chipy_org/apps/job_board/models.py

```diff
@@ -105,11 +106,32 @@ def save(self, *args, **kwargs):  # pylint: disable=arguments-differ
             self.status_change_date = datetime.datetime.now()
             self.__original_status = self.status
 
-        # if post is approved, set the approval date and expiration date
+        # If post is approved and the approval_date hasn't been set yet,
+        # set the approval_date and expiration_date.
         if self.status == "AP" and not self.approval_date:
+
             self.approval_date = datetime.datetime.now()
             self.expiration_date = self.approval_date + datetime.timedelta(days=self.days_to_expire)
 
+            # If days_to_expire is being changed from its initial default value,
+            # it will differ from the __original_days_to_expire field.
+            # If that's the case, change __original_days_to_expire so that in
+            # the future, we can keep track of when the days_to_expire field changes.
+            if self.__original_days_to_expire != self.days_to_expire:
+                self.__original_days_to_expire = self.days_to_expire
+
+        # If post is approved AND expiration_date has been set AND days_to_expire is changed,
+        # then recalculate expiration_date.
+        elif (
+            self.status == "AP"  # pylint: disable=bad-continuation
+            and self.approval_date  # pylint: disable=bad-continuation
+            and self.__original_days_to_expire  # pylint: disable=bad-continuation
+            != self.days_to_expire  # pylint: disable=bad-continuation
+        ):
+
+            self.expiration_date = self.approval_date + datetime.timedelta(days=self.days_to_expire)
+            self.__original_days_to_expire = self.days_to_expire
```

Comment:
I'm having a hard time following this logic. Is it easier to write a nested if block here?

```python
if self.status == "AP":
  ## update expiration date
  if not self.approval_date: #set approval date
  if self.__original_days_to_expire != self.days_to_expire: #set odte
```
     ...

Reply:
@raymondberg Thanks for the comment. Your code isn't equivalent to my code. 
One of the conditions in my code is `if self.approval_date` whereas yours tests for `if NOT self.approval_date`. 
......................................................................................
IN THE BEGINNING 

Here is a hypothetical scenario. After a user creates a post, the status will be `submitted`. When the admin first changes the status to `approved`, the 'if' statement on line 118 will go into effect and the approval_date and expiration_date are set.

```
if self.status == "AP" and not self.approval_date:
    self.approval_date = datetime.datetime.now()
    self.expiration_date = self.approval_date + datetime.timedelta(days=self.days_to_expire)
```
............................................................................................................................
LATER ON

Let's say, later on, the admin wants to extend the job post so that it will show up for more than 60 days. Let's say they want 90 days. The admin will then change the `days_to_expire` field in the admin panel. After the admin presses **save**, the condition for the 'elif' statement on line 132 will now be satisfied.

At this point, the post is in an `approved` state. The approval_date already exists at this point, because we set it the very first time when the post was approved. Since the `days_to_expire` is a new value of **90** and `__original_days_to_expire` is **60**, then they are not equivalent.

(This is the code without the automatic formatting and lint disables.)
```
# If post is approved AND expiration_date has been set AND days_to_expire is changed,
# then recalculate expiration_date.

elif ( self.status == "AP"  and self.approval_date and self.__original_days_to_expire  != self.days_to_expire ):
    self.expiration_date = self.approval_date + datetime.timedelta(days=self.days_to_expire)
    self.__original_days_to_expire = self.days_to_expire
```
We have to recalculate the expiration_date because it depends on the fields 'approval_date' (which is the same value as it was before) and 'days_to_expire' (which value has now changed).  

`__original_days_to_expire` is set to  `self.days_to_expire` to keep track of if `days_to_expire` will change in the future.

I think the code in the actual PR looks confusing because of the automatic formatting that that had to be done and all of the lint disables! :smile:

Reply:
@raymondberg, For readibility's sake, I can turned it into a nested statement. Thanks for your great feedback! :apple:
```
elif self.status == "AP" and self.approval_date:
     
    if self.__original_days_to_expire != self.days_to_expire:
        
        self.expiration_date = self.approval_date + datetime.timedelta(
            days=self.days_to_expire
            )
        
        self.__original_days_to_expire = self.days_to_expire
```

Reply:
Interesting. I think we might need some real good model unit tests here to document what is expected and call out this bug. I'm not sure I understand entirely (which means tests would be great to enforce what you wrote up). But I am curious if something like this would work: https://github.com/chicagopython/chipy.org/compare/master...raymondberg:further-expiration-date-rb?expand=1

Reply:
One reason why my code is complicated is that the expiration_date and days_to_expire field need to be in sync with each other.  I wanted the admin to be able to adjust the number of days as an integer, because it's easy to think in whole numbers about how many days a post should be shown. The expiration_date is then calculated programmatically since humans can't easily calculate dates accurately.  I wanted the computer to be able to use the expiration_date as a filter in the SQL query, where post are shown` if today's date <= expiration date `. I've thought about making `expiration_date` into a property, but you can't filter a Django queryset off of a property.

**I think your code is a great revision.** 

Here is how I would have written it: 
https://gist.github.com/sophm256/93f437a151b632ae47f6d20acbce5fd9

**That being said, I think we should go with your version of the code**. The only difference in your code and mine is that an extra "update_expiration_date()" is run an extra time in your code in the following scenario:

```
Admin sets the status to 'approved' for the first time.
Admin changes the 'days_to_expire'.
Admin then presses 'save' button. 
```

THE DIFFERENCE BETWEEN MY CODE AND YOUR CODE
In my code in that scenario, the update_expiration_date() would only run once overall.  The expiration date is calculated with the correct values the first time under the if statement `if not self.approval_date:` line 119 from my gist. 

In your code, the update_expiration_date runs twice: the first time when it hits `if not self.approval_date:` (line 119) and a second time when it hits `if self.__original_days_to_expire != self.days_to_expire:` (line 127 from your code) . Technically, `update_expiration_date()` doesn't need to run twice, because it was calculated the first time correctly in this particular scenario.

........................................................

I've made a new commit with essentially what you've suggested. **I also have added some more code to account for another condition (line 131 to 136 in my 3rd commit)**. If the admin approves the post but then changes the status to something different, it changes the approval_date and expiration_date back to None. **I will try to write up some tests eventually for this in the future in a separate PR.** Thanks again for your help!

Reply:
Okay I am late to the party, but could you just not calculate an expire date and avoid syncing it.  Just keep the approval date and the number of days to expire value.   I would push the work to view that only displays active posts, i.e approved and approval date + days to expire > datetime.now.  

May make things cleaner.